src/vm-memory/src/lib.rs

```diff
@@ -155,3 +155,205 @@ pub mod test_utils {
+                        std::ptr::write(addr, 0xFF);
+                    },
+                }
+                unreachable!();
+            }
+            child_pid => {
+                let mut child_status: i32 = -1;
+                let pid_done = unsafe { libc::waitpid(child_pid, &mut child_status, 0) };
+                assert_eq!(pid_done, child_pid);
+
+                // Asserts that the child process terminated because
+                // it received a signal that was not handled.
+                assert!(libc::WIFSIGNALED(child_status));
+                // Signal code should be a SIGSEGV
+                assert_eq!(libc::WTERMSIG(child_status), libc::SIGSEGV);
+            }
+        };
+    }
+
+    fn validate_guard_region(region: &MmapRegion) {
+        let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
+
+        // Check that the created range allows us to write inside it
+        let addr = region.as_ptr();
+
+        unsafe {
+            std::ptr::write(addr, 0xFF);
+            assert_eq!(std::ptr::read(addr), 0xFF);
+        }
+
+        // Try a read/write operation against the left guard border of the range
+        let left_border = (addr as usize - page_size) as *mut u8;
+        apply_operation_on_address(left_border, AddrOp::Read);
+        apply_operation_on_address(left_border, AddrOp::Write);
+
+        // Try a read/write operation against the right guard border of the range
+        let right_border = (addr as usize + region.size()) as *mut u8;
+        apply_operation_on_address(right_border, AddrOp::Read);
+        apply_operation_on_address(right_border, AddrOp::Write);
+    }
+
+    fn loop_guard_region_to_sigsegv(region: &MmapRegion) {
+        let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
+
+        let pid = unsafe { libc::fork() };
+        match pid {
+            0 => {
+                let mut addr = region.as_ptr() as usize;
+                let right_page_guard = addr + region.size();
+                loop {
+                    unsafe {
+                        std::ptr::write(addr as *mut u8, 0xFF);
+                    }
+
+                    if addr == right_page_guard {
+                        break;
+                    }
+                    addr += page_size;
+                }
+                unsafe {
+                    libc::exit(0);
+                }
+            }
+            child_pid => {
+                let mut child_status: i32 = -1;
+                let pid_done = unsafe { libc::waitpid(child_pid, &mut child_status, 0) };
+                assert_eq!(pid_done, child_pid);
+
+                // Asserts that the child process terminated because
+                // it received a signal that was not handled.
+                assert!(libc::WIFSIGNALED(child_status));
+                // Signal code should be a SIGSEGV (11)
+                assert_eq!(libc::WTERMSIG(child_status), 11);
+            }
```

Comment:
the logic is indeed pretty similar, but adding a parameter for telling whether it should assert success or sigsegv would not suffice.
This function forks a new process and loops through the entire region.
In contrast, the `apply_operation_on_address()` forks on every call.

Forking for every iteration of the loop should work in theory, but it would require an enormous amount of new processes, and the test would likely timeout.

Reply:
Can't we have the looping logic outside of `apply_operation_on_address()` ? And call `apply_operation_on_address()` just to test if the operation is performed successfully ?

Reply:
the problem is the forking logic, since `apply_operation_on_address()` performs a `fork()` before doing the operation (because it may receive sigsegv).
We can take the forking logic outside `apply_operation_on_address()` and create another helper function like `fork_and_apply_operation` that forks and then calls `apply_operation_on_address`, so that we don't have to add the forking logic in the several places that need it

Reply:
Sounds good to me